### PR TITLE
Supports tx sequence lists

### LINKED ISSUE #993: Add transaction trace output to the json output format

### PR DESCRIPTION
**Pull request by norhh**

### CODE REVIEW

File: mythril/laser/ethereum/instructions.py [hunk 1]
```diff
@@ -1842,6 +1842,14 @@ def call_(self, global_state: GlobalState) -> List[GlobalState]:
             callee_address, callee_account, call_data, value, gas, memory_out_offset, memory_out_size = get_call_parameters(
                 global_state, self.dynamic_loader, True
             )
+
+            if callee_account is not None and callee_account.code.bytecode == "":
+                log.debug("The call is related to ether transfer between accounts")
```
Comment (norhh) [inline mythril/laser/ethereum/instructions.py]: without this, it seems to cause some weird index errors(due to the way cfg is handled) if this isn't placed here. It's a bit hard to place it on the transaction handling area so this might be a quicker way to fix that.

File: mythril/laser/ethereum/svm.py [hunk 2]
```diff
@@ -149,6 +168,11 @@ def sym_exec(
                     "No contract was created during the execution of contract creation "
                     "Increase the resources for creation execution (--max-depth or --create-timeout)"
                 )
+            else:
+                # Replace previous initial state with new ones after contract creation
+                self.set_standard_initial_state(
```
Comment (JoranHonig) [inline mythril/laser/ethereum/svm.py]: Why is this necessary. In this case the initial state would just be the one with the creator and attacker account right? The creation transaction is included in the transaction sequence

File: mythril/laser/ethereum/svm.py [hunk 3]
```diff
@@ -132,10 +148,13 @@ def sym_exec(
 
         if main_address:
             log.info("Starting message call transaction to {}".format(main_address))
+            self.set_standard_initial_state(self.world_state.accounts)
```
Comment (JoranHonig) [inline mythril/laser/ethereum/svm.py]: this can be moved before the if statement, so we don't need the statement on ln 157 as well

File: mythril/laser/ethereum/svm.py [hunk 4]
```diff
@@ -114,6 +115,21 @@ def accounts(self) -> Dict[str, Account]:
         """
         return self.world_state.accounts
 
+    def set_standard_initial_state(
+        self, accounts: Dict[str, Account], ignore_addr=True
+    ):
+        initial_state = self.world_state.initial_state_account
```
Comment (JoranHonig) [inline mythril/laser/ethereum/svm.py]: What's the reason for introducing a new satastructure? Both initial state, and initial_state_account in the world state class

File: mythril/analysis/report.py [hunk 5]
```diff
@@ -64,6 +63,24 @@ def __init__(
         self.source_mapping = None
         self.discovery_time = time() - StartTime().global_start_time
         self.bytecode_hash = get_code_hash(bytecode)
+        if transaction_sequence is None:
+            self.tx_sequence_users = None
```
Comment (JoranHonig) [inline mythril/analysis/report.py]: why are these variables?

File: mythril/analysis/report.py [hunk 6]
```diff
@@ -64,6 +63,24 @@ def __init__(
...
+            self.transaction_sequence_jsonv2 = None
+        else:
+            self.tx_sequence_users = json.dumps(transaction_sequence, indent=4)
+            self.transaction_sequence_jsonv2 = self.add_block_data(transaction_sequence)
+
+    @staticmethod
+    def add_block_data(transaction_sequence: Dict):
```
Comment (JoranHonig) [inline mythril/analysis/report.py]: Shouldn't this be done by the logic that creates the concrete transaction sequence?

File: mythril/analysis/solver.py [hunk 7]
```diff
@@ -109,29 +116,37 @@ def get_transaction_sequence(global_state, constraints):
...
-        concrete_transactions[tx_id]["caller"] = "0x" + (
+        value = model.eval(transaction.call_value.raw, model_completion=True).as_long()
+        concrete_transaction["value"] = "0x%x" % value
+        origin = "0x" + (
             "%x" % model.eval(transaction.caller.raw, model_completion=True).as_long()
         ).zfill(40)
+        concrete_transaction["origin"] = origin
+        concrete_transaction["address"] = "%s" % transaction.callee_account.address
+        concrete_transactions.append(concrete_transaction)
+        min_price_dict[origin] = min_price_dict.get(origin, 0) + value
```
Comment (JoranHonig) [inline mythril/analysis/solver.py]: I don't think origin is necessarily the one who pays the price right? The difference between origin and sender

File: mythril/laser/ethereum/svm.py (hunk 4 above)
Comment (norhh) [reply mythril/laser/ethereum/svm.py]: there is 1 var: initial_state_account, which will just have an initial account state and is propagated everywhere through reference everywhere.

File: mythril/analysis/report.py (hunk 6 above)
Comment (norhh) [reply mythril/analysis/report.py]: But we need to use this from symbolic tx sequence stuff right?

File: mythril/analysis/report.py (hunk 5 above)
Comment (norhh) [reply mythril/analysis/report.py]: tx_sequence_users is used for displaying for users with "<blockTime>" sort of field than some concrete field. (tx_sequence_jsonv2 is for maestro)

File: mythril/laser/ethereum/svm.py (hunk 2 above)
Comment (norhh) [reply mythril/laser/ethereum/svm.py]: Here the initial state would have the created account.

File: mythril/analysis/solver.py (hunk 7 above)
Comment (norhh) [reply mythril/analysis/solver.py]: It's acutally caller which is being used, according to mythril origin can literally be anything.Will rename them.

File: mythril/laser/ethereum/svm.py (hunk 2 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/svm.py]: That doesn't make much sense. Then the creation transaction will create another new account, if you try to replay the transaction sequence?

File: mythril/analysis/report.py (hunk 5 above)
Comment (JoranHonig) [reply mythril/analysis/report.py]: But what's the reason for storing them as class variables versus just calculating when they are displayed

File: mythril/analysis/report.py (hunk 6 above)
Comment (JoranHonig) [reply mythril/analysis/report.py]: What do you mean?

File: mythril/laser/ethereum/svm.py (hunk 2 above)
Comment (norhh) [reply mythril/laser/ethereum/svm.py]: The creation transaction isn't included in the steps I think.

File: mythril/analysis/report.py (hunk 6 above)
Comment (norhh) [reply mythril/analysis/report.py]: We don't have the concrete data of these gasLimit, blockNumber etc right?

File: mythril/analysis/report.py (hunk 5 above)
Comment (norhh) [reply mythril/analysis/report.py]: the function self.add_block_data, and also json.dumps() should be written in multiple places if that's the case, in this case it's just easily reusing them.

File: mythril/analysis/report.py (hunk 6 above)
Comment (JoranHonig) [reply mythril/analysis/report.py]: The blocknumber could be constrained right?

Comment (JoranHonig) [reply mythril/analysis/report.py]: Same for some of the other variables, they are part of the execution environment and can be queried through some of the opcodes

File: mythril/analysis/report.py (hunk 5 above)
Comment (JoranHonig) [reply mythril/analysis/report.py]: I can find only one use of the variables. 
My main point is that the actual data is in transaction_sequence. 
Creating two variables which maintain some representation of the same data doesn't seem like a good approach

I'd either store the initial data, and do the conversion when you need it in some function. 
Or create some functions which do something like
```python
@property
def tx_sequence_users(self):
    return json.dumps(self.transaction_sequence, indent=4) if self.transaction_sequence else None
```

File: mythril/analysis/report.py (hunk 6 above)
Comment (norhh) [reply mythril/analysis/report.py]: yes, but I wanted to do that after this gets merged, there are other things too which I skipped at the moment, I wanted to create a new issue for them to be tackled after the merge. (Like if an issue pops up  in creation tx etc)

Comment (JoranHonig) [reply mythril/analysis/report.py]: Sounds reasonable :+1:

File: mythril/laser/ethereum/instructions.py (hunk 1 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/instructions.py]: With my pr #1025 we might not have this problem anymore

File: mythril/laser/ethereum/svm.py (hunk 2 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/svm.py]: Ok, then this is okay

File: mythril/laser/ethereum/svm.py (hunk 4 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/svm.py]: But depending on the trace through the constructor this account can be different

Comment (JoranHonig) [reply mythril/laser/ethereum/svm.py]: Additionally this introduces some record of accounts using dicts while we already have a datamodel for account's

Comment (JoranHonig) [reply mythril/laser/ethereum/svm.py]: How about we keep track of each transactions' initial and final world state. Then, when serializing you just take the state of the world before the first transaction that you are interested in. 
This is rather simple to add, whereas the current approach adds more complexity to the svm

wdyt

File: mythril/analysis/solver.py [hunk 8]
```diff
@@ -83,19 +90,19 @@ def get_transaction_sequence(global_state, constraints):
 
     transaction_sequence = global_state.world_state.transaction_sequence
 
-    # gaslimit & gasprice don't exist yet
     tx_template = {
```
Comment (JoranHonig) [inline mythril/analysis/solver.py]: can we remove this initialization. There is only one use, so it does not make much sense to initialize this here

File: mythril/laser/ethereum/svm.py (hunk 4 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/svm.py]: it also seems that the transaction serialization is dependent on the structure of the account dicts, this dependency is not apparent and could easily break in the future if we would change or add variables

File: mythril/laser/ethereum/svm.py [hunk 9]
```diff
@@ -114,6 +115,21 @@ def accounts(self) -> Dict[str, Account]:
...
+        initial_state["accounts"] = {}  # This variable persists for all world states.
+        for address, account in accounts.items():
+            if ignore_addr and address == "0x" + "0" * 40:
+                continue
+            initial_state["accounts"][address] = {
+                "nounce": account.nonce,
+                "balance": "<ARBITRARY_BALANCE>",
+                "code": account.code.bytecode,
+                "storage": {},
```
Comment (JoranHonig) [inline mythril/laser/ethereum/svm.py]: An initial state doesn't necessarily not have any storage

Comment (JoranHonig) [reply mythril/laser/ethereum/svm.py]: This would result in incomplete transaction traces right?

File: mythril/laser/ethereum/svm.py (hunk 4 above)
Comment (norhh) [reply mythril/laser/ethereum/svm.py]: `How about we keep track of each transactions' initial and final world state. Then, when serializing you just take the state of the world before the first transaction that you are interested in` But most of the data might redundant right?
`But depending on the trace through the constructor this account can be different`
It's a dict of all accounts(it seems to be a bad variable name). 
`this dependency is not apparent and could easily break in the future if we would change or add variables` If we change variables then yes, else no, I think that should be fine as the members are allowed access there, some tests might help here WDYT?

Comment (norhh) [reply mythril/laser/ethereum/svm.py]: It seems we already store those :man_facepalming:.

File: mythril/laser/ethereum/svm.py [hunk 10]
```diff
@@ -1,9 +1,9 @@
 """This module implements the main symbolic execution engine."""
 import logging
 from collections import defaultdict
-from copy import copy
+from copy import copy, deepcopy
```
Comment (JoranHonig) [inline mythril/laser/ethereum/svm.py]: These are now unused imports right?

File: mythril/analysis/solver.py [hunk 11]
```diff
@@ -109,29 +108,49 @@ def get_transaction_sequence(global_state, constraints):
...
-        concrete_transactions[tx_id]["caller"] = "0x" + (
+        value = model.eval(transaction.call_value.raw, model_completion=True).as_long()
+        concrete_transaction["value"] = "0x%x" % value
+        caller = "0x" + (
             "%x" % model.eval(transaction.caller.raw, model_completion=True).as_long()
         ).zfill(40)
 
-    return concrete_transactions
+        concrete_transaction["origin"] = caller
+        concrete_transaction["address"] = "%s" % transaction.callee_account.address
```
Comment (JoranHonig) [inline mythril/analysis/solver.py]: why do the caller concretization and not

File: mythril/analysis/solver.py [hunk 12]
```diff
@@ -109,29 +108,49 @@ def get_transaction_sequence(global_state, constraints):
...
-        )
-        concrete_transactions[tx_id]["caller"] = "0x" + (
+        value = model.eval(transaction.call_value.raw, model_completion=True).as_long()
+        concrete_transaction["value"] = "0x%x" % value
+        caller = "0x" + (
             "%x" % model.eval(transaction.caller.raw, model_completion=True).as_long()
         ).zfill(40)
 
-    return concrete_transactions
+        concrete_transaction["origin"] = caller
```
Comment (JoranHonig) [inline mythril/analysis/solver.py]: The caller and origin are not the same

File: mythril/analysis/solver.py [hunk 13]
```diff
@@ -109,29 +108,49 @@ def get_transaction_sequence(global_state, constraints):
...
+                initial_state = transaction.world_state.accounts
+                break
+
+    states = {}
+    for address, account in initial_state.items():
+        if address == "0x" + 40 * "0" and len(account.code.bytecode) > 0:
+            continue
+        data = dict()  # type: Dict[str, Union[int, str]]
+        data["nonce"] = account.nonce
+        data["balance"] = account.balance
```
Comment (JoranHonig) [inline mythril/analysis/solver.py]:
```suggestion
        data["balance"] = account.balance()
```

Comment (JoranHonig) [reply mythril/analysis/solver.py]: balance is written twice here

File: mythril/analysis/solver.py [hunk 14]
```diff
@@ -109,29 +108,49 @@ def get_transaction_sequence(global_state, constraints):
...
+    if len(transactions) > 0:
+        initial_state = transactions[0].world_state.accounts
+    else:
+        for transaction in transaction_sequence:
+            if transaction.id == "1":
+                initial_state = transaction.world_state.accounts
+                break
+
+    states = {}
+    for address, account in initial_state.items():
```
Comment (JoranHonig) [inline mythril/analysis/solver.py]: initial_state can be uninitialized

Comment (norhh) [reply mythril/analysis/solver.py]: There will always be a tx_id=“1” right?

File: mythril/analysis/solver.py (hunk 13 above)
Comment (norhh) [reply mythril/analysis/solver.py]: I still didn’t merge it with develop

File: mythril/analysis/solver.py (hunk 12 above)
Comment (norhh) [reply mythril/analysis/solver.py]: We don’t have an origin, so I put in caller there

File: mythril/analysis/solver.py (hunk 14 above)
Comment (JoranHonig) [reply mythril/analysis/solver.py]: Right now, maybe. But we never document/test that this is always the case, so any refactor can change this by accident

Comment (norhh) [reply mythril/analysis/solver.py]: Now it should be fine

File: mythril/analysis/templates/report_as_text.jinja2 [hunk 15]
```diff
@@ -18,11 +18,11 @@ In file: {{ issue.filename }}:{{ issue.lineno }}
 
 --------------------
 {% endif %}
-{% if verbose and issue.debug %}
+{% if verbose and issue.tx_sequence %}
 --------------------
 Transaction Sequence:
 
-{{ issue.debug }}
+{{ issue.tx_sequence }}
```
Comment (JoranHonig) [inline mythril/analysis/templates/report_as_text.jinja2]: Shouldn't this be something like:
```suggestion
{{ issue.tx_sequence_users }}
```

Comment (norhh) [reply mythril/analysis/templates/report_as_text.jinja2]: jinja takesin the issue.as_dict(), where this is tx_sequence

File: mythril/analysis/solver.py [hunk 16]
```diff
@@ -109,29 +108,42 @@ def get_transaction_sequence(global_state, constraints):
...
+        caller = "0x" + (
             "%x" % model.eval(transaction.caller.raw, model_completion=True).as_long()
         ).zfill(40)
 
-    return concrete_transactions
+        concrete_transaction["origin"] = caller
+        concrete_transaction["address"] = "%s" % transaction.callee_account.address
+        concrete_transactions.append(concrete_transaction)
+        min_price_dict[caller] = min_price_dict.get(caller, 0) + value
+    initial_state = transaction_sequence[0].world_state.accounts
```
Comment (JoranHonig) [inline mythril/analysis/solver.py]: If the creation transactions are not included, shouldn't we take the world state after the creation and not before?

Comment (norhh) [reply mythril/analysis/solver.py]: It ran perfectly in my tests, there seems to be a copy stuff not happening somewhere in worldstate.

**Comment by b-mueller** [comment]
Looking good?

File: mythril/analysis/solver.py [hunk 17]
```diff
@@ -83,55 +90,103 @@ def get_transaction_sequence(global_state, constraints):
...
+    # Get concrete values from transaction
+    address = transaction.callee_account.address
+    value = model.eval(transaction.call_value.raw, model_completion=True).as_long()
+    caller = "0x" + (
+        "%x" % model.eval(transaction.caller.raw, model_completion=True).as_long()
+    ).zfill(40)
+
+    # TODO: Do this at Laser
+    if isinstance(transaction, ContractCreationTransaction):
+        calldata = transaction.code.bytecode
```
Comment (JoranHonig) [inline mythril/analysis/solver.py]:
```suggestion
        input = transaction.code.bytecode
```

File: mythril/analysis/solver.py [hunk 18]
```diff
@@ -83,55 +90,103 @@ def get_transaction_sequence(global_state, constraints):
...
+    else:
+        calldata = "".join(
```
Comment (JoranHonig) [inline mythril/analysis/solver.py]:
```suggestion
        input = "".join(
```

File: mythril/analysis/solver.py [hunk 19]
```diff
@@ -83,55 +90,103 @@ def get_transaction_sequence(global_state, constraints):
...
+                hex(b)[2:] if len(hex(b)) % 2 == 0 else "0" + hex(b)[2:]
+                for b in transaction.call_data.concrete(model)
+            ]
+        )
 
-            minimize.append(transaction.call_data.calldatasize)
-            minimize.append(transaction.call_value)
+    # Create concrete transaction dict
+    concrete_transaction = dict()  # type: Dict[str, str]
+    concrete_transaction["input"] = "0x" + calldata
```
Comment (JoranHonig) [inline mythril/analysis/solver.py]:
```suggestion
    concrete_transaction["input"] = "0x" + input_
```

File: mythril/analysis/solver.py [hunk 20]
```diff
@@ -83,55 +90,103 @@ def get_transaction_sequence(global_state, constraints):
...
+    """ Gets a concrete transaction from a transaction and z3 model"""
+    # Get concrete values from transaction
+    address = transaction.callee_account.address
+    value = model.eval(transaction.call_value.raw, model_completion=True).as_long()
+    caller = "0x" + (
+        "%x" % model.eval(transaction.caller.raw, model_completion=True).as_long()
+    ).zfill(40)
+
+    # TODO: Do this at Laser
+    if isinstance(transaction, ContractCreationTransaction):
```
Comment (JoranHonig) [inline mythril/analysis/solver.py]:
```suggestion
    if isinstance(transaction, ContractCreationTransaction):
        address = ""
```

File: mythril/analysis/solver.py [hunk 21]
```diff
@@ -83,55 +90,103 @@ def get_transaction_sequence(global_state, constraints):
...
+def _get_concrete_transaction(model: z3.Model, transaction: BaseTransaction):
+    """ Gets a concrete transaction from a transaction and z3 model"""
+    # Get concrete values from transaction
+    address = transaction.callee_account.address
+    value = model.eval(transaction.call_value.raw, model_completion=True).as_long()
+    caller = "0x" + (
+        "%x" % model.eval(transaction.caller.raw, model_completion=True).as_long()
+    ).zfill(40)
+
+    # TODO: Do this at Laser
```
Comment (JoranHonig) [inline mythril/analysis/solver.py]:
```suggestion
```
